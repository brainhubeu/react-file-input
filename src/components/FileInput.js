import React, { Component } from 'react';
import PropTypes from 'prop-types';
import apsectRatio from 'aspectratio';

import { handleChangeEvent, handleDropEvent, preventDefault } from '../helpers/event';
import { selectIsDragging, selectIsDraggingOver } from '../helpers/fileInputSelectors';

import DropArea from './DropArea';
import FileInputMetadata from './FileInputMetadata';
import ImageThumbnail from './ImageThumbnail';

import '../styles/FileInput.scss';

class FileInput extends Component {
  constructor(props) {
    super(props);

    this.state = {
      enteredInDocument: 0,
      isOver: 0,
      value: null,
    };

    this.input = null;

    this.openFileDialog = this.openFileDialog.bind(this);
    this.selectFile = this.selectFile.bind(this);
    this.getImageThumbnail = this.getImageThumbnail.bind(this);

    this.onDocumentDragEnter = this.onDocumentDragEnter.bind(this);
    this.onDocumentDragLeave = this.onDocumentDragLeave.bind(this);
    this.onDocumentDrop = this.onDocumentDrop.bind(this);

    this.onDragEnter = this.onDragEnter.bind(this);
    this.onDragLeave = this.onDragLeave.bind(this);
    this.onDrop = this.onDrop.bind(this);
  }

  componentDidMount() {
    const { dragOnDocument, dropOnDocument } = this.props;

    if (dragOnDocument) {
      document.addEventListener('dragenter', this.onDocumentDragEnter, false);
      document.addEventListener('dragleave', this.onDocumentDragLeave, false);
      document.addEventListener('drop', this.onDocumentDrop, false);
    }

    if (!dropOnDocument) {
      document.addEventListener('dragover', preventDefault, false);
      document.addEventListener('drop', preventDefault, false);
    }
  }

  componentWillUnmount() {
    const { dragOnDocument, dropOnDocument } = this.props;

    if (dragOnDocument) {
      document.removeEventListener('dragenter', this.onDocumentDragEnter);
      document.removeEventListener('dragleave', this.onDocumentDragLeave);
      document.removeEventListener('drop', this.onDocumentDrop);
    }

    if (!dropOnDocument) {
      document.removeEventListener('dragover', preventDefault);
      document.removeEventListener('drop', preventDefault);
    }
  }

  openFileDialog() {
    if (this.input) {
      this.input.click();
    }
  }

  getImageThumbnail(file = { mimeType: '' }) {
    const reader = new FileReader();
    if (file) {
      if (file.mimeType.match('image.*')) {
        reader.readAsDataURL(file);
      } else {
        this.setState({ image: '' });
      }
    }
    reader.onload = event => {
      const ratio = this.props.scaleImageOptions
        && apsectRatio.crop(
          this.props.scaleImageOptions.width,
          this.props.scaleImageOptions.height,
          this.props.scaleImageOptions.ratio);

      const [width, height] = [ratio[2], ratio[3]];

      const image = this.props.scaleImageOptions
        ? new Image(width, height)
        : new Image();
      image.src = event.target.result;
      image.onload = event => {
        this.setState({ image: event.path[0] });
      };
    };
  }

  selectFile(event) {
    const { onChangeCallback } = this.props;

    const files = handleChangeEvent(event);

    if (files) {
      const file = files[0]; // get only one

      this.setState(state => ({
        ...state,
        enteredInDocument: 0,
        isOver: 0,
        value: file,
      }), () => {
        if (onChangeCallback) {
          onChangeCallback(this.state);
        }
        this.getImageThumbnail(file);
      });
    }
  }

  onDocumentDragEnter(event) {
    this.setState(state => ({ ...state, enteredInDocument: state.enteredInDocument + 1 }));
  }
  onDocumentDragLeave(event) {
    this.setState(state => ({ ...state, enteredInDocument: state.enteredInDocument - 1 }));
  }
  onDocumentDrop(event) {
    this.setState(state => ({ ...state, enteredInDocument: 0 }));
  }

  onDragEnter(event) {
    const { onDragEnterCallback } = this.props;
    const wasDraggingOver = selectIsDraggingOver(this.state);

    this.setState(state => ({ ...state, isOver: state.isOver + 1 }), () => {
      if (onDragEnterCallback && !wasDraggingOver) {
        onDragEnterCallback(this.state);
      }
    });
  }
  onDragLeave(event) {
    const { onDragLeaveCallback } = this.props;
    const wasDraggingOver = selectIsDraggingOver(this.state);

    this.setState(state => ({ ...state, isOver: state.isOver - 1 }), () => {
      if (onDragLeaveCallback && wasDraggingOver) {
        onDragLeaveCallback(this.state);
      }
    });
  }
  onDrop(event) {
    const { onDropCallback } = this.props;

    const files = handleDropEvent(event);

    if (files) {
      const file = files[0]; // get only one

      this.setState(state => ({
        ...state,
        enteredInDocument: 0,
        isOver: 0,
        value: file,
      }), () => {
        if (onDropCallback) {
          onDropCallback(this.state);
        }
        this.getImageThumbnail(file);
      });
    }
  }

  render() {
    const { value, image } = this.state;
    const { customMetadata: CustomMetadata, customImageThumbnail: CustomImageThumbnail, label } = this.props;

    const MetadataClass = CustomMetadata || FileInputMetadata;
    const ImageThumbnailClass = CustomImageThumbnail || ImageThumbnail;

    const isDragging = selectIsDragging(this.state);

    const metadataComponent = value
      && <MetadataClass name={value.filename} size={value.size} extension={value.extension} type={value.mimeType}/>;
    const imageThumbnailComponent = image && <ImageThumbnailClass image={image.src}/>;

    return (
      <div className="brainhub-file-input__wrapper">
        <div className="brainhub-file-input__label">{label}</div>
        <input
          className="brainhub-file-input__input--hidden"
          type="file"
          ref={ref => {
            this.input = ref;
          }}
          onChange={this.selectFile}
        />
        <button onClick={this.openFileDialog}>Select File</button>
        <div className={!isDragging && 'brainhub-file-input__dropInfo--hidden' || ''}>
          <p>Drop here to select file</p>
        </div>
        <div className="brainhub-file-input__fileInfo">
          {this.props.displayImageThumbnail && imageThumbnailComponent}
          {this.props.displayMetadata && metadataComponent}
        </div>
        <DropArea
          dragging={isDragging}
          onDragEnter={this.onDragEnter}
          onDragLeave={this.onDragLeave}
          onDrop={this.onDrop}
          openFileDialog={this.openFileDialog}
        />
      </div>
    );
  }
}

FileInput.defaultProps = {
  dragOnDocument: true,
  dropOnDocument: false,
  onChangeCallback: null,
  onDragEnterCallback: null,
  onDragLeaveCallback: null,
  onDropCallback: null,
  displayMetadata: true,
  displayImageThumbnail: true,
  scaleImageOptions: null,
};

FileInput.propTypes = {
  dragOnDocument: PropTypes.bool,
  dropOnDocument: PropTypes.bool,
  label: PropTypes.string.isRequired,
  onChangeCallback: PropTypes.func,
  onDragEnterCallback: PropTypes.func,
  onDragLeaveCallback: PropTypes.func,
  onDropCallback: PropTypes.func,
  displayMetadata: PropTypes.bool,
  displayImageThumbnail: PropTypes.bool,
  customMetadata: PropTypes.func,
  customImageThumbnail: PropTypes.func,
  scaleImageOptions: PropTypes.shape({
    width: PropTypes.number,
    height: PropTypes.number,
    ratio: PropTypes.string,
  }),
};

export default FileInput;
