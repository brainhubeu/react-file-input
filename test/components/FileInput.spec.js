import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';

import DropArea from '../../src/components/DropArea';
import FileInput from '../../src/components/FileInput';

const defaultProps = {
  label: 'Test',
};

const setup = (props = {}) => {
  const fileInput = mount(<FileInput {...defaultProps} {...props} />);

  return {
    fileInput,
    label: fileInput.find('.brainhub-file-input__label'),
    input: fileInput.find('input[type="file"]'),
    DropArea: fileInput.find(DropArea),
  };
};

const DefaultImage = <img src="image_source" width={500} height={300} />;

const CustomComponent = ({ name='custom component', size=1000, image='custom_image_source' }) =>
  <div>{name}:{size}:{image ? image : DefaultImage}</div>;
CustomComponent.propTypes = {
  name: PropTypes.string,
  size: PropTypes.number,
  image: PropTypes.string,
};

describe('components', () => {
  describe('FileInput', () => {
    const data = {
      name: 'Cute puppies',
      filename: 'Cute puppies',
      extension: 'as',
      size: 1000,
    };

    it('should render with a default state', () => {
      const { fileInput } = setup();

      expect(fileInput.state()).toEqual({
        enteredInDocument: 0,
        isOver: 0,
        value: null,
      });
    });

    it('should render a hidden file input', () => {
      const { input } = setup();
      expect(input).toHaveLength(1);
      expect(input.prop('type')).toBe('file');
      expect(input.hasClass('brainhub-file-input__input--hidden')).toBeTruthy();
    });

    it('should render a label', () => {
      const { label } = setup();

      expect(label).toHaveLength(1);
      expect(label.text()).toBe(defaultProps.label);
    });

    it('should render a DropArea', () => {
      const { DropArea } = setup();

      expect(DropArea).toHaveLength(1);
    });

    it('should render a DropArea that can open the file dialog', () => {
      const { DropArea, input } = setup();

      const inputClickSpy = jest.spyOn(input.instance(), 'click');

      DropArea.prop('openFileDialog')();

      expect(inputClickSpy).toHaveBeenCalled();
    });

    it('should call onDragEnter callback when drag enters for first time', () => {
      const onDragEnterCallback = jest.fn();
      const { DropArea, fileInput } = setup({ onDragEnterCallback });

      DropArea.simulate('dragenter');

      expect(onDragEnterCallback).toHaveBeenCalledWith(fileInput.state());

      onDragEnterCallback.mockClear();
      DropArea.simulate('dragenter');

      expect(onDragEnterCallback).not.toHaveBeenCalled();
    });

    it('should decreate `isOver` state when drag leaves', () => {
      const { DropArea, fileInput } = setup();

      const state = fileInput.state();

      DropArea.simulate('dragLeave');

      expect(fileInput.state()).toEqual({
        ...state,
        isOver: state.isOver - 1,
      });
    });

    it('should call onDragLeaves callback when drag leaves for last time', () => {
      const onDragLeaveCallback = jest.fn();
      const { DropArea, fileInput } = setup({ onDragLeaveCallback });

      fileInput.setState({ isOver: 1 });
      DropArea.simulate('dragLeave');

      expect(onDragLeaveCallback).toHaveBeenCalledWith(fileInput.state());

      onDragLeaveCallback.mockClear();
      DropArea.simulate('dragLeave');

      expect(onDragLeaveCallback).not.toHaveBeenCalled();
    });

    it('should store the drop file in the state, and reset the rest, and call the callback', () => {
      const onDropCallback = jest.fn();
      const { DropArea, fileInput } = setup({ onDropCallback });

      const file = { name: 'MockFile.mkv', size: 5000 };
      fileInput.setState({ isOver: 2, enteredInDocument: 2, image: '' });

      DropArea.simulate('drop', { dataTransfer: { files: [file] } });

      expect(fileInput.state('enteredInDocument')).toBe(0);
      expect(fileInput.state('isOver')).toBe(0);
      expect(fileInput.state('value')).toBe(file);
      expect(onDropCallback).toHaveBeenCalledWith(fileInput.state());
    });

    it('should render metadata from default component', () => {
      const { fileInput } = setup();
      fileInput.setState({ value: data });

      expect(fileInput.find('FileInputMetadata')).toHaveLength(1);
    });

    it('should render metadata from custom component', () => {
      const { fileInput } = setup({ customMetadata: CustomComponent });
      fileInput.setState({ value: data });

      expect(fileInput.find(CustomComponent)).toBeTruthy();
      expect(fileInput.find('FileInputMetadata')).toHaveLength(0);
    });

    it('should not render metadata when user pass false to displayMetadata prop', () => {
      const { fileInput } = setup({ displayMetadata: false });
      fileInput.setState({ value: data });

      expect(fileInput.find('FileInputMetadata').length).toEqual(0);
    });

    it('should render image thumbnail from default component', () => {
      const { fileInput } = setup();
      fileInput.setState({ image: 'image_source' });

      expect(fileInput.find('ImageThumbnail')).toHaveLength(1);
    });

    it('should render image thumbnail from custom component', () => {
      const { fileInput } = setup({ customImageThumbnail: CustomComponent });
      fileInput.setState({ image: 'image_source' });

      expect(fileInput.find(CustomComponent)).toBeTruthy();
      expect(fileInput.find('ImageThumbnail')).toHaveLength(0);
    });

    it('should not render mage thumbnail when user pass false to displayImageThumbnail prop', () => {
      const { fileInput } = setup({ displayImageThumbnail: false });
      fileInput.setState({ value: data });

      expect(fileInput.find('ImageThumbnail').length).toEqual(0);
    });

    it('should render image in defined ratio when scaleImageOptions prop were pass', () => {
      const scaleImageOptions = { width: 100, height: 100, ratio: '16:9' };
      const { input, fileInput } = setup({ scaleImageOptions });
      const data = {
      //   name: 'Cute puppies',
      //   filename: 'Cute puppies',
      //   extension: 'as',
      //   size: 1000,
      // };
      // const image = <img width={500} height={300} />;
      // fileInput.setState({ image });

      // input.simulate('change', {
      //   target: {
      //     files: [
      //       data,
      //     ],
      //   },
      // });
      // // expect(fileInput.find('ImageThumbnail'))
      // console.log(fileInput.state());
    });

    it('should match exact snapshot', () => {
      const tree = renderer.create(
        <div>
          <FileInput {...defaultProps} />
        </div>
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
