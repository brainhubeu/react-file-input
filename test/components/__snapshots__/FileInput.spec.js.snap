// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components FileInput should match exact snapshot 1`] = `
<div>
  <div
    className="brainhub-file-input__wrapper"
  >
    <div
      className="brainhub-file-input__label"
    >
      Test
    </div>
    <input
      className="brainhub-file-input__input--hidden"
      onChange={[Function]}
      type="file"
    />
    <div
      className="brainhub-file-input__fileInfo"
    />
    <div
      className="brainhub-drop-area"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDrop={[Function]}
    >
      <div
        className="brainhub-drop-area__button"
        onClick={[Function]}
        role="button"
        tabIndex={-1}
      >
        + Add File
      </div>
    </div>
  </div>
</div>
`;
